src/server/dcp_server/models.py

```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
+        
+        ## TODO implement the eval pipeline, i.e. first call self.segmentor.eval, then split again into patches
+        ## using resulting seg and then call self.classifier.eval. The final mask which is returned should have 
+        ## first channel the output of cellpose and the rest are the class channels
+       
+
+        final_mask = self.get_prediction(img, instance_mask)
+        # instance segmentation mask (C, W, H) --> semantic multiclass segmentation mask (W, H)
+        for i in range(instance_mask.shape[0]):
+            instance_mask[i, ...][instance_mask[i, ...] > 0] = i + 1
+
+        seg_mask = instance_mask.sum(0)
+
+        final_mask = torch.tensor(final_mask).to(int)
+        seg_mask = torch.tensor(seg_mask)
+        
+
+        jaccard_score = JaccardIndex(task="multiclass", num_classes=4, ignore_index=0)
+
+        jaccard_index = jaccard_score(final_mask.sum(0), seg_mask)
+        
+
+        return final_mask, jaccard_index
+
+    def get_prediction(self, input_image, cellpose_mask):
+        """
+        Performs object segmentation and classification on an input image using the Cellpose model and a classifier model.
+
+        Args:
+            image_path (str): The file path of the input image.
+            model (CellposeModel): The Cellpose model used for object segmentation, instance segmenation mask.
+         
+
+        Returns:
+            tuple: A tuple containing the cellpose_mask and final_mask, representing the segmentation masks obtained from
+                the Cellpose model and the combined segmentation and classification mask, respectively.
+        """
+
+        # Obtain segmentation mask using Cellpose model
+
+        # Find objects in the cellpose_mask
+        locs = find_objects(cellpose_mask)
+
+        # Get patches and labels based on object centroids
+        patches, labels = self.get_centered_patches(input_image, cellpose_mask, int(1.5 * input_image.shape[0] // 5), noise_intensity=5)
+        
+        labels = torch.tensor(labels)
+        labels_fit = []
+        
+        final_mask = torch.zeros(cellpose_mask.shape)
+
+        with torch.no_grad():
+            for i, patch in enumerate(patches):
+                loc = locs[i]
+
+                # Prepare image patch for classification
+                img = torch.tensor(patch.astype(np.float32)).unsqueeze(0).unsqueeze(0) / 255
+                # img = img.mean(dim=1, keepdim=True)
+              
+                # Perform inference using model_classifier
+                logits = self.classifier(img) 
+        
+                _, predicted = torch.max(logits, 1)
+                labels_fit.append(predicted)
+
+                # Assign predicted class to corresponding location in final_mask
+                final_mask[loc] = predicted + 1
+
+        # Apply mask to final_mask, retaining only regions where cellpose_mask is greater than 0
+        final_mask = final_mask * ((cellpose_mask > 0).long())
```

Comment:
why do we need this? won't locs give you coordinates of non-zero object locations?

Reply:
locs are the slices, hence final_mask[loc] = x, will always fill an entire region (rectangle given by slice) with the value of x. To fill out the empty regions multiply by cellpose mask.